### PR TITLE
supports multiple NCCL communicators preserved in NCCLCommContext

### PR DESCRIPTION
**Pull request by gavin1332**
supports multiple NCCL communicators preserved in NCCLCommContext.
This pr adapt paddle pipeline training for single-process-and-multiple-threads training
test=develop

### CODE REVIEW

File: paddle/fluid/platform/collective_helper.h [hunk 1]
```diff
@@ -58,37 +58,57 @@ class NCCLComm {
...
+  void CreateAllNCCLComms(const std::vector<int>& dev_ids, int ring_id = 0);
+
+  // retrieve a communicator by the ring id in multiprocessing mode
   NCCLComm* Get(int ring_id) const {
-    PADDLE_ENFORCE(comm_map_.count(ring_id),
-                   "comunicator in ring id %d has not been initialized",
-                   ring_id);
-    return comm_map_.at(ring_id).get();
+    PADDLE_ENFORCE_GT(comm_map_.count(ring_id), 0,
+                      "comunicator in ring id %d has not been initialized",
```
Comment (hutuxian) [inline paddle/fluid/platform/collective_helper.h:77]: comunicator -> communicator

File: paddle/fluid/platform/collective_helper.h [hunk 2]
```diff
@@ -58,37 +58,57 @@ class NCCLComm {
...
+                      ring_id);
+    PADDLE_ENFORCE_GT(
+        comm_map_.at(ring_id).count(dev_id), 0,
+        "comunicator at device id %d has not been initialized in ring %d",
+        dev_id, ring_id);
+    return comm_map_.at(ring_id).at(dev_id).get();
+  }
+
+  // retrieve a communicator by the ring id and place
+  NCCLComm* Get(int ring_id, Place place) const {
```
Comment (hutuxian) [inline paddle/fluid/platform/collective_helper.h:98]: const Place& place?

**Review by hutuxian — APPROVED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]
LGTM for paddle_enforce

File: paddle/fluid/platform/collective_helper.h (hunk 1 above)
Comment (gavin1332) [reply paddle/fluid/platform/collective_helper.h:77]: leave it to TODO list

File: paddle/fluid/platform/collective_helper.h (hunk 2 above)
Comment (gavin1332) [reply paddle/fluid/platform/collective_helper.h:98]: the place is a POD, not necessary.

**Pull request merged**